controllers/users_config_map_controller.go

```diff
@@ -0,0 +1,274 @@
+		}
+	}
+
+	// reread quay token
+	r.QuayClient = r.BuildQuayClient(log)
+
+	// remove team if config map was removed, renamed, or doesn't contain any users
+	if removeTeam {
+		log.Info("Will remove team", "TeamName", teamName)
+		if err := r.QuayClient.DeleteTeam(r.QuayOrganization, teamName); err != nil {
+			log.Error(err, "failed to remove team", "TeamName", teamName, l.Action, l.ActionDelete)
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// fetch image repositories in the namespace
+	imageRepositoryList := &imagerepositoryv1alpha1.ImageRepositoryList{}
+	if err := r.Client.List(ctx, imageRepositoryList, &client.ListOptions{Namespace: configMap.Namespace}); err != nil {
+		log.Error(err, "failed to list ImageRepositories", l.Action, l.ActionView)
+		return ctrl.Result{}, err
+	}
+
+	// get image repositories names
+	repositoriesToAdd := []string{}
+	for _, repository := range imageRepositoryList.Items {
+		repositoriesToAdd = append(repositoriesToAdd, repository.Spec.Image.Name)
+	}
+
+	// get team members, if team doesn't exist it will be created
+	log.Info("Ensure team", "TeamName", teamName)
+	teamMembers, err := r.QuayClient.EnsureTeam(r.QuayOrganization, teamName)
+	if err != nil {
+		log.Error(err, "failed to get team members", "TeamName", teamName, l.Action, l.ActionView)
+		return ctrl.Result{}, err
+	}
+	// get users in the team
+	usersInTeam := []string{}
+	for _, user := range teamMembers {
+		usersInTeam = append(usersInTeam, user.Name)
+	}
+	log.Info("Users in the team", "TeamName", teamName, "Users", usersInTeam)
+
+	// get team permissions
+	teamPermissions, err := r.QuayClient.ListPermissionsForTeam(r.QuayOrganization, teamName)
+	if err != nil {
+		log.Error(err, "failed to get team permissions", "TeamName", teamName, l.Action, l.ActionView)
+		return ctrl.Result{}, err
+	}
+	// get repositories for which team has permissions
+	reposTeamPermissions := []string{}
+	for _, repoPermission := range teamPermissions {
+		reposTeamPermissions = append(reposTeamPermissions, repoPermission.Repository.Name)
+	}
+	log.Info("Team has repository permissions", "TeamName", teamName, "Repositories", reposTeamPermissions)
+
+	// grant repo permissions to the team
+	for _, repoToUpdate := range repositoriesToAdd {
+		shouldGrantPermission := true
+		for _, repoWithPermission := range reposTeamPermissions {
+			if repoToUpdate == repoWithPermission {
+				shouldGrantPermission = false
+				break
+			}
+		}
+
+		if shouldGrantPermission {
+			log.Info("Grant repository permission to the team", "TeamName", teamName, "RepositoryName", repoToUpdate)
+			if err := r.QuayClient.AddReadPermissionsForRepositoryToTeam(r.QuayOrganization, repoToUpdate, teamName); err != nil {
+				log.Error(err, "failed to grant repo permission to the team", "TeamName", teamName, "RepositoryName", repoToUpdate, l.Action, l.ActionAdd)
+				return ctrl.Result{}, err
+			}
+		}
+	}
+
+	// add users to the team
```

Comment:
Does it make sense to have a function that would return usersToAdd and usersToRemove lists and then the operations will be applied to the ready lists? (will be slightly easier to read the code).

Reply:
added helper functions